mdf_matio/validator.py

```diff
@@ -0,0 +1,397 @@
+
+        # BLOCK: files
+        # Add file data to dataset
+        if rc_md["files"]:
+            self.__indexed_files += rc_md["files"]
+
+        # BLOCK: material
+        # elements
+        if rc_md["material"].get("composition"):
+            composition = rc_md["material"]["composition"].replace("and", "")
+            # Currently deprecated
+#                for element in DICT_OF_ALL_ELEMENTS.keys():
+#                    composition = re.sub("(?i)"+element,
+#                                         DICT_OF_ALL_ELEMENTS[element], composition)
+            str_of_elem = ""
+            for char in list(composition):
+                if char.isupper():  # Start of new element symbol
+                    str_of_elem += " " + char
+                elif char.islower():  # Continuation of symbol
+                    str_of_elem += char
+                # Anything else is not an element (numbers, whitespace, etc.)
+
+            # Split elements in string (on whitespace), make unique and JSON-serializable
+            list_of_elem = list(set(str_of_elem.split()))
+            # Ensure deterministic results
+            list_of_elem.sort()
+            # Currently deprecated
+            # If any "element" isn't in the periodic table,
+            # the composition is likely not a chemical formula and should not be parsed
+#                if all([elem in DICT_OF_ALL_ELEMENTS.values() for elem in list_of_elem]):
+#                    record["elements"] = list_of_elem
+
+            rc_md["material"]["elements"] = list_of_elem
+        elif rc_md["material"].get("elemental_proportions"):
+            rc_md["material"]["elements"] = [rc_md["material"]["elemental_proportions"].keys()]
+            rc_md["material"]["elements"].sort()
+
+        # BLOCK: custom
+        # Make all values into strings
+        if rc_md.get("custom"):
+            new_custom = {}
+            for key, val in rc_md["custom"].items():
+                new_custom[key] = str(val)
+            rc_md["custom"] = new_custom
+
+        # Require strict JSON
+        try:
+            json.dumps(rc_md, allow_nan=False)
+        except (ValueError, json.JSONDecodeError) as e:
+            return {
+                "success": False,
+                "error": "Invalid record JSON: {}".format(str(e)),
+                "details": repr(e)
+                }
+
+        # Remove null/None values
+        rc_md = _remove_nulls(rc_md, self.__allowed_nulls)
+
+        # Validate against schema
+        try:
+            jsonschema.validate(rc_md, schema, resolver=resolver)
+        except jsonschema.ValidationError as e:
+            return {
+                "success": False,
+                "error": "Invalid record metadata: " + str(e).split("\n")[0],
+                "details": str(e)
+                }
+
+        # Write out to file
+        json.dump(rc_md, self.__tempfile)
+        self.__tempfile.write("\n")
```

Comment:
If we do not perform any further computation using a record after this point, why store it at all? Could we just return the validated records and let the user of the class decide if it is needed to store them or whether they can be passed out of the generator.

Reply:
[Discussed in main thread]